### PR TITLE
Updating doc building to Sphinx 4.3.2 (python 3)

### PR DESCRIPTION
**Pull request by barsch**
This PR updates the docs build process to Sphinx 3.1.1. After unsucessfully trying to fix the old doc builder I decided to restart the whole process from scratch and reimplement existing features afterwards.

- Bootstrap theme got replaced by more lightweight "Read the Docs" theme - however I didn't spend to much time on styling yet as this is something which should be tackled after the build process is running smooth
- stylesheet/javascript/image optimization is done via gulp
- played a bit with sphinx-gallery but didn't like it, so its again a modified version of matplotlib.sphinxext.plot_directive, which work good enough for me
- all plots using basemap or mlpy are failing -> segmentation fault on ubuntu as well on win10 for basemap 1.2.1 on py38
- docs were in pretty bad shape - a lot of warnings have been fixed - there are still many left
- currently only the html builder has been tested - latex/docset etc need work
- the structure of the start page probably needs some tweak - I suggest to remove all help/mailing list options from there and link to a github page instead - we should focus on documentation only IMHO

A preview of the current results can be seen here: https://docs.obspy.org/sphinx3/

Feel free to participate - running the doc building takes unfortunatly ages ... **If someone could rewrite all examples using basemap to cartopy would be a big help**.

### CODE REVIEW

**Comment by megies** [comment]
Awesome stuff @barsch. I like the overall looks, and I'm real happy that we'll have a proper navigation again (with the left side bar).

Can you put the exact version numbers of the build environment somewhere, I can help set up the cron jobs later.. or.. we could have @d-chambers and @heavelock give a shot at building docs in gh-actions? Might be the right time to give it a try, since we would have to set up docs building on new server from scratch more or less anyway?

Also, can you put the build log as a gist maybe.. interesting to see if it got any more readable at all compared to the 80-page warning/error logs we used to have

**Comment by barsch** [comment]
@megies its already build on the new server - log can be found here http://docs-test.obspy.org/log.txt

**Comment by megies** [comment]
OK, oof, still such a bookshelf of warnings.

**Comment by barsch** [comment]
jupp - and still takes ages to run - its not fun working on it ...

**Comment by megies** [comment]
No kidding, it might easily be the most painful thing to work on in the whole project.

**Comment by adigitoleo** [comment]
Hey I was browsing the docs and saw that there are some bits where sphinx links to the builtin ``time`` module, just because ``time`` is used as a parameter name:



It's a very small thing and it seems like you've got enough work cut out already, but I thought I'd point it out since you're updating the docs. It might be solved using the `:samp:` sphinx role, but I've never used it:

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-samp

**Comment by heavelock** [comment]
Hey! I had some spare time today and I thought about finally working on building docs on GithubActions. Coming from GitlabCI this action systems seems to be weird but I think I managed to make the builder run.

Check it out on PR against my fork:
https://github.com/heavelock/obspy/pull/9

It's based on obspy/sphinx3 branch that was rebased on top of current master

**Comment by heavelock** [comment]
Oh dammit. I accidentally pushed to wrong remote and my changes got in here. Sorry for that as I meant to first ask before doing that ...

Anyway, should I remove my changes or can they stay here?

My changes include:
1.  Rebasing on top of current master
2.  Add GH action for building in CI
3. GH action for exproting the artifact with built docs (not yet fully done as there are some problems with exact path that it built to)
4. Added cartopy to dependencies of docs so we can start fixing plotting routines.
5. Pinned sphinx==3.4.3 as 3.5.0 that was released yesterday has a bug that prevents from building.

**Comment by barsch** [comment]
All fine - knock yourself out ;)

**Comment by heavelock** [comment]
Since we are at the docs, do you think it would beneficial for us to start using some linter for docs?
In my projects I'm using doc8 which is equivalent to flake8 for rst docs.

What do you think?

**Comment by heavelock** [comment]
Basemap -> Cartopy conversion progress:
- [x] basemap_with_beachball_read_events.py
- [ ] basemap_plot_with_beachballs2.py
- [ ] basemap_plot_with_beachballs.py
- [ ] basemap_plot_custom.py

**Comment by barsch** [comment]
@heavelock first of all big thanks for looking into this! Two rants from my side while you are on it:

* can you also look into this: https://github.com/obspy/obspy/pull/2662#issuecomment-723529200
* important for me would be to have always the possibility to run the docs build process locally in case gh-actions fails to work/they charge money in the future or for whatever reason

**Comment by heavelock** [comment]
1. Yeah, sure. Not today though, I have to come back to my normal tasks for now. I will take a look on that tomorrow.

And the second point:
Definitely. How about having a dedicated docker image for building it?

**Comment by barsch** [comment]
docker would be perfect imho

**Comment by heavelock** [comment]
perf. I had it already done but then I ditched it because the sphinx-action I am using here does not support using a custom image as argument. I could fork it and have our own action for that or just create a custom workflow. Positive side of the action is that it is parsing the warnings from sphinx and dropping them in diff.

**Comment by heavelock** [comment]
I testes the `:samp:` directive but it breaks the rendering. I noticed that this problem occurs only for class variables, it is not cross-referenced for standard `:param:`. I guesss it might be a bug upstream. I will post it to the sphinx's bugtracker.
obraz

**Comment by heavelock** [comment]
@barsch Would you mind forking this repo as organization? https://github.com/ammaraskar/sphinx-action
It is the action that I used for building docs. Unfortunately, I don't see a possibility of hanging a docker image used by it without modifying the code.

**Comment by barsch** [comment]
> @barsch Would you mind forking this repo as organization? https://github.com/ammaraskar/sphinx-action

done: https://github.com/obspy/sphinx-action

**Comment by ThomasLecocq** [comment]
my turn to give it a go :-)

**Comment by ThomasLecocq** [comment]
**2022 changes**

- Rebased on current master
- Changed the template for `class` so that the _Private_ and _Special_ methods don't create a new page, but are listed/inline on the Class's page. This results in a gigantic reduction of output html (there was one for each of ``__eq__``, ``__del__`` etc, for each class. It makes longer pages, but all the info (and cross-refs) are preserved: see example [after/before]:
image

- Disabled rendering hires & pdf for plots, it takes a lot of time for no real usefulness
- removed the ":numbered:" param to the tutorial
- The build on github actions + @heavelock 's docker takes ~20 min, and the artifact works smooth

on my local machine, the first build took also ~20 minutes, successive builds (adding one comma to one RST file): less than 2 minutes. This means, running a doc builder that doesn't redo everything from scratch _could_ produce docs in a flash. That'd require having some workflow like:

**first run ever (ever)**
- create conda env (+cache it)
- pull master branch
- build the doc
- upload the whole package content as artifact

- activate doc environment
- download the artifact
- pull/checkout the branch
- build doc
- export .build/html folder somewhere


TODO still:
- [ ] revert or adapt the changelog.rst vs changelog.md
- [ ] try to shrink the 3554 (!) warnings
- [ ] find a way to reduce slow processes:

```
====================== slowest reading durations =======================
452.730 packages/autogen/obspy.imaging.cm
42.885 packages/autogen/obspy.core.inventory.network.Network.plot
34.954 tutorial/code_snippets/beamforming_fk_analysis
29.142 packages/autogen/obspy.core.event.Catalog.plot
9.388 tutorial/code_snippets/waveform_plotting_tutorial
```

**Comment by trichter** [comment]
For me the images in the gallery are a bit large.

> doesn't list Event

Also no Netwrok, Station, Channel, but that is intentional?

**Comment by ThomasLecocq** [comment]
> > Also no Network, Station, Channel, but that is intentional?
> 
> No, that's linked to Inventory's `:no-index:` too, on it :-)

solved the browsing + added class names to the info of packages, so they are also references & clickable
image

**Comment by barsch** [comment]
> > How could we make this work together with docs.obspy.org?

I can take care of it - alternatively you can get an SSH account and do it yourself

**Comment by ThomasLecocq** [comment]
> > > How could we make this work together with docs.obspy.org?
> 
> I can take care of it - alternatively you can get an SSH account and do it yourself

Hi @barsch : I have zero docker knowledge so if this is the way to go, I'm maybe not very helpful. But yes, for other reasons (incl. debugging), I'm happy to SSH to the machine & pull the release / master docs & build them

**Comment by barsch** [comment]
> > > > How could we make this work together with docs.obspy.org?
> > 
> > 
> > I can take care of it - alternatively you can get an SSH account and do it yourself
> 
> Hi @barsch : I have zero docker knowledge so if this is the way to go, I'm maybe not very helpful. But yes, for other reasons (incl. debugging), I'm happy to SSH to the machine & pull the release / master docs & build them

I know docker only as user and can't help developing it either -> thats where @heavelock comes in

however the last master docs (https://docs.obspy.org/master) were build using this branch without docker so SSH makes sense - I'll contact you privately later (working atm)

**Comment by ThomasLecocq** [comment]
>however the last master docs (https://docs.obspy.org/master) were build using this branch without docker so SSH makes sense - I'll contact you privately later (working atm)

roger that, sounds like a plan!

**Comment by trichter** [comment]
Ideally this could imo work as previously (when the docs were still working), i.e. upload docs to https://docs.obspy.org/pr/$PRnumber if `build_docs` tag is present and automatically update https://docs.obspy.org/master/ on each new commit in master.
Finally upload to  https://docs.obspy.org for a new tag.

We also need to update the instructions how to build docs locally.
https://github.com/obspy/obspy/wiki/Sphinx-documentation

**Comment by trichter** [comment]
@ThomasLecocq You replaced several images, by reason or accident?

**Comment by ThomasLecocq** [comment]
> @ThomasLecocq You replaced several images, by reason or accident?

that was before me :-)

**Comment by ThomasLecocq** [comment]
> Ideally this could imo work as previously (when the docs were still working), i.e. upload docs to https://docs.obspy.org/pr/$PRnumber if `build_docs` tag is present and automatically update https://docs.obspy.org/master/ on each new commit in master. Finally upload to https://docs.obspy.org for a new tag.
> 
> We also need to update the instructions how to build docs locally. https://github.com/obspy/obspy/wiki/Sphinx-documentation

agree with all this. Just dunno how to "deploy" safely to a remote server (from the gh-actions)

**Comment by barsch** [comment]
> > Ideally this could imo work as previously (when the docs were still working), i.e. upload docs to https://docs.obspy.org/pr/$PRnumber if `build_docs` tag is present and automatically update https://docs.obspy.org/master/ on each new commit in master. Finally upload to https://docs.obspy.org for a new tag.
> > We also need to update the instructions how to build docs locally. https://github.com/obspy/obspy/wiki/Sphinx-documentation
> 
> agree with all this. Just dunno how to "deploy" safely to a remote server (from the gh-actions)

the old "magic" was done here: https://github.com/obspy/obspy/blob/master/misc/scripts/docs_buildbot/update-docs-pr.sh

**Comment by trichter** [comment]
> agree with all this. Just dunno how to "deploy" safely to a remote server (from the gh-actions)

rsync via ssh?
Copy to destination folders with rsync and touch a file pr/$pr-number.touch.

A server-side script can then delete old pr docs.

**Comment by ThomasLecocq** [comment]
> 

seems doable indeed:
https://github.com/marketplace/actions/ssh-deploy

And the good thing is that now it's quite easy to have any of _us_ run the doc builder on a recent conda env, in case GHactions become limited or expensive

File: misc/docs/source/packages/obspy.io.nied.rst [hunk 1]
```diff
@@ -1,7 +1,9 @@
-:orphan:
+:oprhan:
```
Comment (ThomasLecocq) [inline misc/docs/source/packages/obspy.io.nied.rst:1]: orphan

**Comment by ThomasLecocq** [comment]
I re-parsed all changes, should be OK now.

**Comment by ThomasLecocq** [comment]
Just saying: it is verrrryyyyy satisfying to work on a PR for many days, feeling "ready" and then seeing someone taking over the next steps 😃 🎈 🥳 @trichter

**Comment by ThomasLecocq** [comment]
forgot to check: might be useful to save the warning log (it's nicely sorted, better than redirecting stdout+stderr (but those are also useful):

On my machines, I run:
` sphinx-build -b html -j 25 -w warnings.log source .build`
`-j 25` is pseudo parallel stuff, using 25 "threads" to _write_ (only!), default is `-j auto` which takes N as N cores on the machine.

I didn't adapt the make scripts yet

**Comment by trichter** [comment]
@ThomasLecocq I was just checking if this deployment stuff is working out of the box. It is not. The container does not have ssh installed. We can either create a new container with ssh, rsync, curl. We could load the artefact in another job. Or we could right away install the dependencies with conda or mamba.

**Comment by ThomasLecocq** [comment]
> @ThomasLecocq I was just checking if this deployment stuff is working out of the box. It is not. The container does not have ssh installed. We can either create a new container with ssh, rsync, curl. We could load the artefact in another job. Or we could right away install the dependencies with conda or mamba.

maybe indeed easier just to use the conda+cache+etc that we master now... :-)

**Comment by trichter** [comment]
> maybe indeed easier just to use the conda+cache+etc that we master now... :-)

> one pinned requirement I forgot to add in misc/docs/requirements.txt is mistune<1 (currently have mistune 0.8.4), as 1+ has an compat issue

Sorry, would have been better to create a new PR and take this branch as base. Please go ahead, I am finished for today.

**Comment by ThomasLecocq** [comment]
> > maybe indeed easier just to use the conda+cache+etc that we master now... :-)
> 
> > one pinned requirement I forgot to add in misc/docs/requirements.txt is mistune<1 (currently have mistune 0.8.4), as 1+ has an compat issue
> 
> Sorry, would have been better to create a new PR and take this branch as base. Please go ahead, I am finished for today.

Roger that (but I'm done for today too :-) , will do later 🍻

**Comment by trichter** [comment]
I reverted my changes here and created a new branch at PR #2947.

**Comment by ThomasLecocq** [comment]
work in progress to make the event & inventory cartopy-proof (might put that in a separate PR, not sure yet):
image

**Comment by trichter** [comment]
> might put that in a separate PR, not sure yet

What fits best for you. I skimmed through changes introduced by this PR and I didn't find any obvious glitches. Maybe we can try to get rid of the image replacements of `_images/*.png` to not grow the size of the repository unnecessarily. Some stuff I don't understand (e.g. gulpfile.js), so maybe @barsch can also have look after you finished.

**Comment by barsch** [comment]
```gulp``` (https://gulpjs.com/docs/en/getting-started/quick-start) is just a build tool to automate CSS, JavaScript and static image management, e.g. minifying code, optimizing images, preprocessing CSS, etc.

I wouldn't try to automate this via GH actions if not needed as this has failed in the past after updating dependencies etc. Instead I suggest we just leave the the ```gulpfile.js``` in the root and once some theme changes are needed one just runs the script locally and creates new static minimized and compressed .css and .js files

**Comment by trichter** [comment]
I checked ObsPy's sectret tab and:

> Secrets are not passed to workflows that are triggered by a pull request from a fork.

Because we want to see doc builds also from forks, it will not work to deploy the doc via SSH in a run triggered in the doc building workflow. We either need a separate github action checking in a time interval for artifacts to upload. Or straightforward, the server can check for runs and relevant artifacts on its own. So the doc building would still be on github actions and can be maintained in this repository, but the server fetches and deploys the docs.

**Comment by barsch** [comment]
> Or straightforward, the server can check for runs and relevant artifacts on its own. So the doc building would still be on github actions and can be maintained in this repository, but the server fetches and deploys the docs.

I very much prefer that.

**Comment by trichter** [comment]
Great. @barsch Do you want to write the necessary server-side code? I could also do it. This PR can serve as a blue-print.

**Comment by ThomasLecocq** [comment]
> Great. @barsch Do you want to write the necessary server-side code? I could also do it. This PR can serve as test load.

@trichter I propose we do it you & me when the PR is ready. I'm almost done with the final plots.

**Comment by barsch** [comment]
there should be code which could be recycled here: https://github.com/obspy/obspy/tree/master/misc/scripts/docs_buildbot

> Great. @barsch Do you want to write the necessary server-side code? I could also do it. This PR can serve as test load.

I'm happy to maintain/admin the server and obspy-reporter, but as stated before coding for obspy is not really on my agenda any more - didn't use obspy since 10 years or so and its about time other take over ;)

(but I'm actually very impressed and quite happy to see huge progress here)

**Comment by ThomasLecocq** [comment]
all look OK now (didn't yet port all the "resolution" stuff, but as a PoC it works) 

image
image
image
image

**Comment by ThomasLecocq** [comment]
YESSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS http://msnoise.org/obspy-doc/packages/autogen/obspy.core.event.Event.plot.html#obspy.core.event.Event.plot

**Comment by barsch** [comment]
hmm

http://msnoise.org/obspy-doc/gallery.html

vs.

http://msnoise.org/obspy-doc/gallery.html

for whatever reason custom.min.css is not used

**Comment by ThomasLecocq** [comment]
> hmm
> 
> http://msnoise.org/obspy-doc/gallery.html
> 
> vs.
> 
> http://msnoise.org/obspy-doc/gallery.html
> 
> for whatever reason custom.min.css is not used

ah, indeed, that... checking now- difference is that on your site div has class "gallery" and not on "

**Comment by ThomasLecocq** [comment]
CI was 🟢 , which is very satisfactory :-) 
now rebuilding the doc locally to update http://msnoise.org/obspy-doc/gallery.html & then I think it'll be good to merge.

And move on with, on the DOC side: the doc builder and on the code side: remove all basemap stuff

**Review by trichter — APPROVED** [review]

**Comment by ThomasLecocq** [comment]
249 commits +2,927 −3,259 ... wow

**Pull request merged**